daemon/containerd/image_tag.go

```diff
@@ -1,20 +1,57 @@
 package containerd
 
 import (
-	"errors"
+	"context"
 
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
 	"github.com/docker/distribution/reference"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/sirupsen/logrus"
 )
 
-// TagImage creates the tag specified by newTag, pointing to the image named
-// imageName (alternatively, imageName can also be an image ID).
-func (i *ImageService) TagImage(imageName, repository, tag string) (string, error) {
-	return "", errdefs.NotImplemented(errors.New("not implemented"))
-}
+// TagImageWithReference creates an image named as newTag and targeting the given descriptor id.
+func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.ID, newTag reference.Named) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"imageID": imageID.String(),
+		"tag":     newTag.String(),
+	})
+
+	target, err := i.resolveDescriptor(ctx, imageID.String())
+	if err != nil {
+		logger.WithError(err).Debug("failed to resolve image id to a descriptor")
+		return err
+	}
+
+	img := containerdimages.Image{
+		Name:   newTag.String(),
+		Target: target,
+	}
+
+	is := i.client.ImageService()
+	_, err = is.Create(ctx, img)
+	if err != nil {
+		if !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Debug("failed to create image")
+			return errdefs.System(err)
```

Comment:
Do we know what errors this can return, and if they should always be considered an internal server error? (does `is.Create()` return typed errors from containerd, and can any of those be a "user error" (4xx)?

(Perhaps a 5xx is okay in all cases, but might be worth checking and/or a comment to outline that irregardless of the error type, we consider it an internal server error).

Reply:
Yeah definitely there are some errors which shouldn't be considered a server error, like: missing name, labels too big. Those are the minority of all errors that could be returned though. The rest look system related (grpc, internal storage failures).
I think we should keep this as `System`, but validate the image ourselves before calling `is.Create()` and return an user error.

Reply:
Actually, in this case already have a valid reference which is a valid image name. So it's not possible for this one to fail with user user. Will change the `Debug` to `Error` log then.